app/crud/serviceMetric.py

```diff
@@ -1,19 +1,142 @@
+from typing import Optional
+from datetime import datetime
 
 class CRUDServiceMetric(CRUDBase[ServiceMetric, ServiceMetricCreate, ServiceMetricUpdate]):
     def __init__(self, db_session: Session):
         super(CRUDServiceMetric, self).__init__(ServiceMetric, db_session)
+        self.scorecardMetrics = CRUDScoreCardMetric(db_session)
+        self.Metric = CRUDMetric(db_session)
 
     def getByScorecardId(self, scorecardId: int) -> list[ServiceMetric]:
         return self.db_session.query(ServiceMetric).filter(ServiceMetric.scorecardId == scorecardId).all()
 
     def getByServiceId(self, serviceId: int) -> list[ServiceMetric]:
         return self.db_session.query(ServiceMetric).filter(ServiceMetric.serviceId == serviceId).all()
 
+    def get_metrics_by_scorecard_and_service(self, scorecard_id: int, service_id: int) -> list[ServiceMetric]:
+        metrics = self.db_session.query(ServiceMetric).filter(
+            ServiceMetric.serviceId == service_id,
+            scoreCardMetrics.scoreCardId == scorecard_id
+        ).all()
+        return metrics
+    
+    # fun for get metric list with range time
+    def get_metric_values_by_service(self, service_id: int,from_date: Optional[datetime], to_date: Optional[datetime]) -> list[ServiceMetricCL]:
+        metrics = (
+            self.db_session.query(ServiceMetric.metricId,ServiceMetric.value,ServiceMetric.timestamp)
+        )
+        query = self.db_session.query(ServiceMetric.metricId, ServiceMetric.value, ServiceMetric.timestamp)  # Query initialized here
+        if from_date and to_date:
+          query = query.filter(ServiceMetric.timestamp >= from_date, ServiceMetric.timestamp <= to_date)
+        elif from_date:
+          query = query.filter(ServiceMetric.timestamp >= from_date)
+        elif to_date:
+          query = query.filter(ServiceMetric.timestamp <= to_date)
+        
+        query = query.filter(ServiceMetric.serviceId == service_id).order_by(ServiceMetric.timestamp.desc())
+        metrics = query.all()
+        return metrics
+     
+     
+      
+    def get_last_metrics(self, scorecard_id: int, service_id: int)-> list[ServiceMetric]:
+        
+        subquery = (
+        self.db_session.query(
+            ServiceMetric.metricId,
+            func.max(ServiceMetric.timestamp).label('latest_timestamp')
+        )
+        .filter(
+            ServiceMetric.serviceId == service_id,
+            ServiceMetric.metricId.in_(self.scorecardMetrics.getMetricByScoreCradId(scorecard_id))
+        )
+        .group_by(ServiceMetric.metricId)
+        .subquery()
+    )
+
+        latest_metrics = self.db_session.query(
+        ServiceMetric.metricId,
+        ServiceMetric.value,
+        ServiceMetric.timestamp
+     ).join(
+        subquery,
+        and_(
+            ServiceMetric.metricId == subquery.c.metricId,
+            ServiceMetric.timestamp == subquery.c.latest_timestamp
+        )
+     ).filter(
```

Comment:
no need for this filter as it's already applied on the subquery.

Reply:
done